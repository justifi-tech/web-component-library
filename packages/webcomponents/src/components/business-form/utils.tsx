import { h } from '@stencil/core';

export const FormAlert = ( message: string ) => (
  <div class='alert alert-danger' role='alert'>
    {message}
  </div>
);
