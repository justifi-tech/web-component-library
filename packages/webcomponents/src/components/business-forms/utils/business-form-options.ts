import { BusinessClassification } from "../../../api/Business";

interface BusinessFormOption {
  label: string;
  value: string;
}

export const businessClassificationOptions: BusinessFormOption[] = [
  {
    label: 'Choose business classification',
    value: '',
  },
  {
    label: 'Sole Proprietor',
    value: BusinessClassification.sole_proprietor,
  },
  {
    label: 'Partnership',
    value: BusinessClassification.partnership,
  },
  {
    label: 'Corporation',
    value: BusinessClassification.corporation,
  },
  {
    label: 'Public Company',
    value: BusinessClassification.public_company,
  },
  {
    label: 'Limited',
    value: BusinessClassification.limited,
  },
  {
    label: 'Non Profit',
    value: BusinessClassification.non_profit,
  },
  {
    label: 'Government',
    value: BusinessClassification.government,
  },
];

export const businessServiceReceivedOptions: BusinessFormOption[] = [
  {
    label: 'Within 7 days',
    value: 'Within 7 days',
  },
  {
    label: 'Within 30 days',
    value: 'Within 30 days',
  },
  {
    label: 'Within 90 days',
    value: 'Within 90 days',
  },
  {
    label: 'Within 120 days',
    value: 'Within 120 days',
  },
  {
    label: 'More than 120 days',
    value: 'More than 120 days',
  }
];

export const recurringPaymentsOptions: BusinessFormOption[] = [
  {
    label: 'Yes',
    value: 'Yes',
  },
  {
    label: 'No',
    value: 'No',
  }
];

export const seasonalBusinessOptions: BusinessFormOption[] = [
  {
    label: 'Yes. The majority of the business revenue is generated in 3-6 months.',
    value: 'Yes',
  },
  {
    label: 'No. The business revenue is generated evenly throughout the year.',
    value: 'No',
  }
];

export const bankAccountTypeOptions: BusinessFormOption[] = [
  {
    label: 'Checking',
    value: 'checking',
  },
  {
    label: 'Savings',
    value: 'savings',
  }
];