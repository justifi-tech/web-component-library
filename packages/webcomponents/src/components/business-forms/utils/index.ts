export * from './business-form-options';
export * from './helpers';
export * from './input-handlers';
export * from './event-types';
export * from './country-form-helpers';
