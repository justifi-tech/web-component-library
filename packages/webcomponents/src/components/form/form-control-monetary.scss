@import 'reboot';
@import 'forms';

:host {
  @include host-base-styles;
  display: block;
}

input[type='text'][disabled] {
  color: var(--jfi-form-control-disabled-color);
  background-color: var(--jfi-form-control-disabled-background-color);
  opacity: 0.5;
  cursor: not-allowed;
}

input {
  border-left: 0 !important;
  border-top-left-radius: 0 !important;
  border-bottom-left-radius: 0 !important;
}
