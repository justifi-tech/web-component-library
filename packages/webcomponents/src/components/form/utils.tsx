import { h } from '@stencil/core';

export const LoadingSpinner = () => (
  <div class="spinner-border spinner-border-sm" role="status">
    <span class="visually-hidden">Loading...</span>
  </div>
);

export const numberOnlyHandler = (e: KeyboardEvent | ClipboardEvent) => {
  const specialKeys = ['Backspace', 'Tab', 'Enter', 'ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight', 'Delete'];

  // Check if the key is a special key
  if (e instanceof KeyboardEvent && specialKeys.includes(e.key)) {
    return;
  }

  // Allow Ctrl+V / Command+V for pasting
  const isPaste = e instanceof KeyboardEvent && (e.ctrlKey || e.metaKey) && e.key.toLowerCase() === 'v';

  if (isPaste) {
    return;
  }

  // Handle paste event
  if (e instanceof ClipboardEvent) {
    const clipboardData = e.clipboardData?.getData('text');
    if (clipboardData && !/^\d+$/.test(clipboardData)) {
      e.preventDefault();
    }
    return;
  }

  // Prevent default action if the key is not a digit
  if (e instanceof KeyboardEvent && !/^\d$/.test(e.key)) {
    e.preventDefault();
  }
};
