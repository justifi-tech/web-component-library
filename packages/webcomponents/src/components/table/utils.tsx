import { h } from '@stencil/core';

export const LoadingState = (columnSpan: number) => (
  <tr>
    <td class="loading-state" part="loading-state-cell" colSpan={columnSpan} style={{ textAlign: 'center' }}>
      <div part="loading-state-spinner" class="spinner-border" role="status">
        <span class="visually-hidden">Loading...</span>
      </div>
    </td>
  </tr>
);

export const EmptyState = (columnSpan: number) => (
  <tr>
    <td class="empty-state" part="empty-state" colSpan={columnSpan} style={{ textAlign: 'center' }}>No results</td>
  </tr>
);

export const ErrorState = (columnSpan: number, errorMessage: string) => (
  <tr>
    <td class="error-state" part="error-state" colSpan={columnSpan} data-test-id="table-error-state" style={{ textAlign: 'center' }}>
      An unexpected error occurred: {errorMessage}
    </td>
  </tr>
);
