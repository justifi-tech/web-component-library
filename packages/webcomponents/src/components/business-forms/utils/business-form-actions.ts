import { ComponentErrorSeverity } from "../../../api/ComponentError";
import { getErrorCode, getErrorMessage } from "../../../api/services/utils";

export const makeGetBusiness = ({ authToken, businessId, service }) =>
  async ({ onSuccess, onError }) => {
    try {
      const response = await service.fetchBusiness(authToken, businessId);

      if (!response.error) {
        const business = response.data;
        onSuccess({ business });
      } else {
        const responseError = getErrorMessage(response.error);
        const code = getErrorCode(response.error?.code);
        return onError({
          error: responseError,
          code,
          severity: ComponentErrorSeverity.ERROR,
        });
      }
    } catch (error) {
      const code = getErrorCode(error?.code);
      return onError({
        error: error.message || error,
        code,
        severity: ComponentErrorSeverity.ERROR,
      });
    }
  };

export const makePatchBusiness = ({ authToken, businessId, service }) =>
  async ({ business, onSuccess, onError }) => {
    try {
      const response = await service.patchBusiness(authToken, businessId, business);

      if (!response.error) {
        onSuccess(response);
      } else {
        const responseError = getErrorMessage(response.error);
        const code = getErrorCode(response.error?.code);
        return onError({
          error: responseError,
          code,
          severity: ComponentErrorSeverity.ERROR,
        });
      }
    } catch (error) {
      const code = getErrorCode(error?.code);
      return onError({
        error: error.message || error,
        code,
        severity: ComponentErrorSeverity.ERROR,
      });
    }
  };
  