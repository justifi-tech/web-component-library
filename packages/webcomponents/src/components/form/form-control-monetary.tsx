import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Watch,
} from '@stencil/core';
import IMask, { InputMask } from 'imask';
import { CURRENCY_MASK } from '../../utils/form-input-masks';

@Component({
  tag: 'form-control-monetary',
  styleUrl: 'form-control-number.scss',
  shadow: true,
})
export class MonetaryInput {
  @Prop() label: string;
  @Prop() name: string;
  @Prop() error: string;
  @Prop() defaultValue: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() maskOptions: any = CURRENCY_MASK.DECIMAL;

  private imask: InputMask<any> | null = null;

  textInput!: HTMLInputElement;

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  componentDidLoad() {
    this.initializeIMask();
    this.updateInput(this.defaultValue);
  }

  disconnectedCallback() {
    this.imask?.destroy();
  }

  private initializeIMask() {
    if (!this.textInput) return;

    this.imask = IMask(this.textInput, this.maskOptions);

    this.imask.on('accept', () => {
      const rawValue = this.imask.unmaskedValue;
      this.inputHandler(this.name, rawValue);
      this.formControlInput.emit({ name: this.name, value: rawValue });
    });

    this.textInput.addEventListener('blur', () => this.formControlBlur.emit());
  }

  updateInput(newValue: any) {
    if (this.imask) {
      this.imask.value = String(newValue);
    }
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <label part="label" class="form-label" htmlFor={this.name}>
          {this.label}
        </label>
        <input
          ref={el => (this.textInput = el as HTMLInputElement)}
          id={this.name}
          name={this.name}
          onBlur={() => this.formControlBlur.emit()}
          part={`input ${this.error && 'input-invalid'}`}
          class={this.error ? 'form-control is-invalid' : 'form-control'}
          type="text"
        />
        {this.error && <div class="invalid-feedback">{this.error}</div>}
      </Host>
    );
  }
}
