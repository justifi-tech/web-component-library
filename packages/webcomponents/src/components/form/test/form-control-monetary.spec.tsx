import { newSpecPage } from "@stencil/core/testing";
import { MonetaryInput } from "../form-control-monetary";
import { CURRENCY_MASK } from "../../../utils/form-input-masks";

describe('form-control-monetary', () => {
  it('renders correctly with default props', async () => {
    const page = await newSpecPage({
      components: [MonetaryInput],
      html: `<form-control-monetary></form-control-monetary>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('handles props correctly', async () => {
    const page = await newSpecPage({
      components: [MonetaryInput],
      html: `
      <form-control-monetary
        label="Amount"
        name="amount"
        error="Invalid amount"  
        defaultValue="1000"
        maskOptions=${CURRENCY_MASK.DECIMAL}
      ></form-control-monetary>
    `,
    });
    await page.waitForChanges();

    const label = page.root.shadowRoot.querySelector('label');
    const input = page.root.shadowRoot.querySelector('input');
    const errorDiv = page.root.shadowRoot.querySelector('.invalid-feedback');

    expect(label.textContent).toBe('Amount');
    expect(input.getAttribute('name')).toBe('amount');
    expect(errorDiv.textContent).toBe('Invalid amount');
  });

  it('calls inputHandler and emits formControlInput on user input', async () => {
    const inputHandlerMock = jest.fn();
    const page = await newSpecPage({
      components: [MonetaryInput],
      html: `<form-control-monetary maskOptions=${CURRENCY_MASK.DECIMAL}></form-control-monetary>`,
    });

    page.rootInstance.inputHandler = inputHandlerMock;
    await page.waitForChanges();

    const inputSpy = jest.fn();
    page.win.addEventListener('formControlInput', inputSpy);

    const input = page.root.shadowRoot.querySelector('input');
    input.value = '1234.56';
    input.dispatchEvent(new Event('input', { bubbles: true }));
    await page.waitForChanges();

    expect(inputHandlerMock).toHaveBeenCalled();
    expect(inputSpy).toHaveBeenCalled();
  });

  it('emits formControlBlur on input blur', async () => {
    const page = await newSpecPage({
      components: [MonetaryInput],
      html: `<form-control-monetary maskOptions=${CURRENCY_MASK.DECIMAL}></form-control-monetary>`,
    });

    const blurSpy = jest.fn();
    page.win.addEventListener('formControlBlur', blurSpy);

    const input = page.root.shadowRoot.querySelector('input');
    input.dispatchEvent(new Event('blur', { bubbles: true }));
    await page.waitForChanges();

    expect(blurSpy).toHaveBeenCalled();
  });

  it('displays error message when error prop is set', async () => {
    const page = await newSpecPage({
      components: [MonetaryInput],
      html: `<form-control-monetary error="Invalid amount" maskOptions=${CURRENCY_MASK.DECIMAL}></form-control-monetary>`,
    });

    const errorDiv = page.root.shadowRoot.querySelector('.invalid-feedback');
    expect(errorDiv.textContent).toBe('Invalid amount');
    expect(page.root).toMatchSnapshot();
  });
});
