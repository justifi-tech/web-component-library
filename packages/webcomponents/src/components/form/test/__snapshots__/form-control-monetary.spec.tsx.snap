// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form-control-monetary displays error message when error prop is set 1`] = `
<form-control-monetary error="Invalid amount" exportparts="label,input,input-invalid" maskoptions="[object" object]="">
  <mock:shadow-root>
    <label class="form-label" part="label"></label>
    <input class="form-control is-invalid" part="input input-invalid" type="text">
    <div class="invalid-feedback">
      Invalid amount
    </div>
  </mock:shadow-root>
</form-control-monetary>
`;

exports[`form-control-monetary renders correctly with default props 1`] = `
<form-control-monetary exportparts="label,input,input-invalid">
  <mock:shadow-root>
    <label class="form-label" part="label"></label>
    <input class="form-control" part="input undefined" type="text">
  </mock:shadow-root>
</form-control-monetary>
`;
